File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/RuntimeFact.cs
namespace TimeOnChrome.AppRuntime.Core;

public enum UserActivityState
{
    Unknown,
    Active,
    Idle,
}

public enum UserSessionState
{
    Unknown,
    Active,
    Inactive,
    Locked,
}

public enum SystemPowerState
{
    Unknown,
    Awake,
    Asleep,
}

public enum RuntimeFactKind
{
    ApplicationActivated,
    UserActivityChanged,
    SessionChanged,
    PowerChanged,
    Snapshot,
}

public sealed record RuntimeSnapshot(
    ApplicationIdentity? Application,
    UserActivityState UserActivity,
    UserSessionState SessionState,
    SystemPowerState PowerState);

public sealed record RuntimeFact
{
    public required long ObservedAtMs { get; init; }

    public required RuntimeFactKind Kind { get; init; }

    public ApplicationIdentity? Application { get; init; }

    public UserActivityState? UserActivity { get; init; }

    public UserSessionState? SessionState { get; init; }

    public SystemPowerState? PowerState { get; init; }

    public RuntimeSnapshot? Snapshot { get; init; }
}

File: macos-app-management/Sources/MacOSAppRuntimeCore/RuntimeFact.swift
public enum UserActivityState: String, Codable, Equatable, Sendable {
    case unknown
    case active
    case idle
}

public enum UserSessionState: String, Codable, Equatable, Sendable {
    case unknown
    case active
    case inactive
    case locked
}

public enum SystemPowerState: String, Codable, Equatable, Sendable {
    case unknown
    case awake
    case asleep
}

public struct RuntimeSnapshot: Codable, Equatable, Sendable {
    public let application: ApplicationIdentity?
    public let userActivity: UserActivityState
    public let sessionState: UserSessionState
    public let powerState: SystemPowerState

    public init(
        application: ApplicationIdentity?,
        userActivity: UserActivityState,
        sessionState: UserSessionState,
        powerState: SystemPowerState
    ) {
        self.application = application
        self.userActivity = userActivity
        self.sessionState = sessionState
        self.powerState = powerState
    }
}

public struct RuntimeFact: Codable, Equatable, Sendable {
    public enum Kind: Codable, Equatable, Sendable {
        case applicationActivated(ApplicationIdentity?)
        case userActivityChanged(UserActivityState)
        case sessionChanged(UserSessionState)
        case powerChanged(SystemPowerState)
        case snapshot(RuntimeSnapshot)
    }

    public let observedAtMs: Int64
    public let kind: Kind

    public init(observedAtMs: Int64, kind: Kind) {
        self.observedAtMs = observedAtMs
        self.kind = kind
    }
}
